### PR TITLE
Clamp default num_threads to cpuset/affinity mask

### LINKED ISSUE #193859: Torch crashes during import/thread-trashes CPU when lauched in cpuset environment
### 🐛 Describe the bug

Easiest way to validate it is to launch something like that
```
bash -c 'ulimit -Su $(( $(ps -eLo user= | grep -c "^$USER") + 60 )); taskset -c 0-7 python -c "import torch; t = torch.ones(50_000_000); torch.equal(t, t)"'
```
on host with lots of CPU cores, which will crash with (ulimit shenanigans are needed to create soft limit on total number of processes that could be launched by given user, which is a pretty typical setup for some Kubernetes deployments, which a much higher limit though) 
```
libgomp: Thread creation failed: Resource temporarily unavailable
```

### Versions

2.13/nightly

cc @ezyang @gchanan @kadeng @msaroufim

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #194605

When no explicit thread count is given, TaskThreadPoolBase::defaultNumThreads() derived the default from cpuinfo, which reports the host's topology (physical cores / processors) and ignores the calling process' CPU affinity mask. In a cpuset/taskset-restricted environment (e.g. a Kubernetes pod pinned to a subset
of cores) this makes PyTorch spawn as many threads as the host has cores, which is inefficient and under a tight ulimit on processes this fails at import with:
```
    libgomp: Thread creation failed: Resource temporarily unavailable
```

Fixes #193859

Test Plan:
Added test_cpuset_num_threads() to .ci/pytorch/test.sh (wired into the default
Linux test branch), which pins the process to a single CPU and asserts the
default thread count collapses to 1:

```
env -u OMP_NUM_THREADS -u MKL_NUM_THREADS taskset -c 0 python -c \
  'import torch; n = torch.get_num_threads(); assert n == 1, n'
```

Co-Authored-By: Claude Opus 4.8 <noreply@anthropic.com>

### CODE REVIEW

**Review by atalman — APPROVED** [review]

**Review by janeyx99 — APPROVED** [review]
oh cool! this is something we wanted for a bit
